### PR TITLE
Allow authenticating dapps to disable PIN auth

### PR DESCRIPTION
**Pull request by nmattia**
This extends the II authentication API (postMessage interface) to include a new boolean field `allowPinAuthentication`. When set to false (default being true) Internet Identity will not allow the user to authenticating to the dapp using a PIN identity, and will not allow the user to register using a PIN identity.

This field is marked as experimental in the spec for now.

### CODE REVIEW

File: demos/test-app/src/index.tsx [hunk 1]
```diff
@@ -203,11 +206,16 @@ const init = async () => {
     const derivationOrigin =
       derivationOriginEl.value !== "" ? derivationOriginEl.value : undefined;
 
+    let allowPinAuthentication = allowPinAuthenticationEl.checked
```
Comment (nmattia) [inline demos/test-app/src/index.tsx:209]: use `const`

File: src/frontend/src/components/displayError.ts [hunk 2]
```diff
@@ -31,7 +33,11 @@ ${options.detail}</pre
       })}
 
       <div class="l-stack">
-        <button id="displayErrorPrimary" class="c-button c-button--primary">
+        <button
+          id="displayErrorPrimary"
+          data-error-code=${ifDefined(options.errorCode)}
```
Comment (lmuntaner) [inline src/frontend/src/components/displayError.ts:38]: Is this related to disabling the PIN?

File: src/frontend/src/components/authenticateBox.ts [hunk 3]
```diff
@@ -699,6 +720,7 @@ const useIdentityFlow = async <T, I>({
   | UnknownUser
   | ApiError
   | BadPin
+  | { kind: "pinNotAllowed" }
```
Comment (lmuntaner) [inline src/frontend/src/components/authenticateBox.ts:723]: Why not created a new type like `BadPin`?

File: src/frontend/src/test-e2e/flows.ts [hunk 4]
```diff
@@ -104,19 +104,51 @@ export const FLOWS = {
...
+      pinAuthView.waitForDisplay().then(() => "pin_allowed" as const),
+      browser
+        .$('#errorContainer [data-error-code="pinNotAllowed"]')
+        .waitForDisplayed()
+        .then(() => "pin_disallowed" as const),
+    ]);
+    if (result === "pin_disallowed") {
+      return "pin_disallowed";
+    }
+    result satisfies "pin_allowed";
```
Comment (nmattia) [inline src/frontend/src/test-e2e/flows.ts:128]: Add a comment to explain what's happening

File: src/frontend/src/components/authenticateBox.ts [hunk 5]
```diff
@@ -734,6 +756,11 @@ const useIdentityFlow = async <T, I>({
   }
   isValid satisfies "valid";
 
+  // if there is a PIN but allowPinAuth is false, then error out
```
Comment (lmuntaner) [inline src/frontend/src/components/authenticateBox.ts:759]: Where is the "PIN" in this function? Is it that `pinIdentityMaterial` is not nullish and `isValid` has `"valid"`?

Comment (lmuntaner) [reply src/frontend/src/components/authenticateBox.ts:759]: Could we also have checked this before checking the `isValid`?

**Review by lmuntaner — APPROVED** [review]
Thanks for sharing this!

I added some comments, but LGTM!

File: src/frontend/src/components/authenticateBox.ts (hunk 3 above)
Comment (nmattia) [reply src/frontend/src/components/authenticateBox.ts:723]: Since they're all just type aliases it doesn't matter too much. Would you prefer having `PinNotAllowed`?

Comment (lmuntaner) [reply src/frontend/src/components/authenticateBox.ts:723]: No, I wanted to understand the patterns you use in the code.

File: src/frontend/src/components/authenticateBox.ts (hunk 5 above)
Comment (nmattia) [reply src/frontend/src/components/authenticateBox.ts:759]: The flow first checks whether or not there's stored PIN material. If there's some, then it means the user is using a PIN and not webauthn. So if the PIN material exists (and is valid, i.e. not expired) then we check if PIN auth is allowed. It would not make sense to check if PIN auth is allowed because it's only relevant in case there PIN material exists.

Does this make sense?

File: src/frontend/src/components/displayError.ts (hunk 2 above)
Comment (nmattia) [reply src/frontend/src/components/displayError.ts:38]: It's related to the tests for disabling the PIN; I wanted to check that the error message shown was indeed that PIN auth is not allowed

File: src/frontend/src/components/authenticateBox.ts (hunk 5 above)
Comment (lmuntaner) [reply src/frontend/src/components/authenticateBox.ts:759]: Ah, yes. Sorry, I got confused because I saw `isNullish(pinIdentityMaterial)` above, and I thought that was all the check needed.

I see that's just one exit in case there is no `pinIdentityMaterial`. But the user might still use a passkey even with a pin material.

**Review by peterpeterparker — APPROVED** [review]
LGTM, thx

**Comment by marydwyer** [comment]
If it meets the request from @peterpeterparker, then it LGTM! Thanks!
